src/write/encoder_string_writer.rs

```diff
@@ -0,0 +1,107 @@
+use crate::Config;
+use std::io;
+use std::io::Write;
+use super::encoder::EncoderWriter;
+
+/// A `Write` implementation that base64-encodes data using the provided config and accumulates the
+/// resulting base64 in memory, which is then exposed as a String via `finish()`.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+///
+/// let mut enc = base64::write::EncoderStringWriter::new(base64::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // get the resulting String
+/// let b64_string = enc.finish().unwrap();
+///
+/// assert_eq!("YXNkZg==", &b64_string);
+/// ```
+///
+/// # Panics
+///
+/// Calling `write()` (or related methods) or `finish()` after `finish()` has completed without
+/// error is invalid and will panic.
+///
+/// # Performance
+///
+/// B64-encoded data is buffered in the heap since the point is to collect it in a String.
+pub struct EncoderStringWriter {
+    encoder: EncoderWriter<Vec<u8>>,
+}
+
+impl EncoderStringWriter {
+    /// Create a new EncoderStringWriter that will encode with the provided config.
+    pub fn new(config: Config) -> EncoderStringWriter {
+        EncoderStringWriter { encoder: EncoderWriter::new(Vec::new(), config) }
+    }
+
+    /// Encode all remaining buffered data, including any trailing incomplete input triples and
+    /// associated padding.
+    ///
+    /// Once this succeeds, no further writes or calls to this method are allowed.
+    ///
+    /// Returns the base64-encoded form of the accumulated written data.
+    ///
+    /// # Errors
+    ///
+    /// The first error that is not of `ErrorKind::Interrupted` will be returned.
+    pub fn finish(&mut self) -> io::Result<String> {
+        let buf = self.encoder.finish()?;
+
+        let str = String::from_utf8(buf).expect("Base64 should always be valid UTF-8");
```

Comment:
Call me risky but I would use `String::from_utf8_unchecked()` and add a `debug_assert!()` that is is valid. There is no reason to check it at runtime when we know that we just generated valid utf8.

Reply:
I just benchmarked to see, and it's only about a 5% performance difference until inputs get very large, at which point I assume people aren't typically buffering 10MiB of data on the heap, but if they are, they'll probably still be ok with 1.7GiB/s encoding instead of 1.9 :)

Reply:
Also note that microbenchmarks aren't always the best. In this case unicode parsing (at least the subset done by `&str`) can take up valuable instruction cache space. However your stance is completely valid so I will resolve this discussion. It can also always be changed later.

Reply:
Point well taken. This library used to use a _lot_ of unsafe and I did get it all out eventually, so I'm hesitant to lose my `forbid(unsafe_code)`. It does pain me to lose the performance, but eventually in some bizarro universe where I have time to do it, SIMD will more than make up for the UTF-8 cost.